### PR TITLE
Use Groovy instead of ant to detect platform/arch

### LINKED ISSUE #761: go cli binary not being generated for mac
## Environment details:

Mac El Capitan

## Steps to reproduce the issue:

1.   git clone latest using hash bf409c1aabc0087cbca0a65ae19683b1d07437f1
2.   Not using vagrant using native mac dev
3.   Run $ gradle :tools:go-cli:distDocker

## Provide the expected results and outputs:

The binary in `/../../bin/go-cli/wsk` is built for mac

## Provide the actual results and outputs:

The binary in `/../../bin/go-cli/wsk` is built for linux

fyi @mdeuser @dubeejw

### PR DESCRIPTION
**Pull request by mdeuser**
Use Groovy instead of ant to detect platform/arch
Fixes #761

### CODE REVIEW

**Comment by mdeuser** [comment]
@dubeejw - Ready for review.  Thanks!

**Comment by mdeuser** [comment]
@csantanapr - Would you mind verifying that this PR fixes #761.  The fix is based off information for Jason's mac, but would like to see how it works on your system.  Thanks!

File: tools/go-cli/build.gradle [hunk 1]
```diff
@@ -78,3 +44,34 @@ def run(cmd, ignoreError = false) {
...
+    if (osname_l.contains("mac")) return "mac"
+    if (osname_l.contains("windows")) return "windows"
+    return osname_l
+}
+
+def mapOsArchToGoArch(String osarch) {
+    String osarch_l = osarch.toLowerCase()
+    if (osarch_l == "amd64") return "amd64"
+    if (osarch_l.contains("x86_64")) return "amd64"
+    if (osarch_l.contains("i386")) return "386"
```
Comment (dubeejw) [inline tools/go-cli/build.gradle]: Should also handle the case where osarch is x86_32, and return 386.

Comment (mdeuser) [inline tools/go-cli/build.gradle]: Sure.  Can add that.  FYI.  This logic in this build.gradle could be short lived as it's probably better to write some sort of run-time wsk wrapper that launches the appropriate wsk binary on-the-fly.  Only for openwhisk development environment though (i.e. someone building openwhisk in vagrant and want to easily run the wsk CLI from vagrant and the host system), not for CLI distribution.

**Comment by dubeejw** [comment]
:+1:

File: tools/go-cli/build.gradle [hunk 2]
```diff
@@ -54,15 +19,16 @@ task distBinary(dependsOn: [removeBinary, distDocker]) << {
     run(["mkdir", "${projectDir}/../../bin/go-cli/"])
     run(dockerBinary + ["run", "--name", dockerContainerName, dockerImageName + ":" + dockerImageTag])
 
-    // Copy all Go binaries outside of Docker into openwhisk go-cli/bin/ folder
+    // Copy all Go binaries outside of Docker into openwhisk go-cli/build/ folder
     run(dockerBinary + ["cp", dockerContainerName +
```
Comment (markusthoemmes) [inline tools/go-cli/build.gradle]: This doesn't copy to the `build` folder as stated by the comment?

File: tools/go-cli/build.gradle [hunk 3]
```diff
@@ -54,15 +19,16 @@ task distBinary(dependsOn: [removeBinary, distDocker]) << {
...
-        exec {
-            executable "ln"
-            args "${projectDir}/../../bin/go-cli/${ant.properties.family}/${ant.properties.arch}/wsk", "${projectDir}/../../bin/go-cli/wsk", "-s"
-        }
+    // Create a shorthand sym link to current platform appropriate Go CLI binary
+    String go_osname = mapOsNameToGoName(getOsName())
+    String go_osarch = mapOsArchToGoArch(getOsArch())
+    exec {
+        executable "ln"
+        args "${projectDir}/../../bin/go-cli/"+go_osname+"/"+go_osarch+"/wsk", "${projectDir}/../../bin/go-cli/wsk", "-s"
```
Comment (markusthoemmes) [inline tools/go-cli/build.gradle]: Use String templating as for `${projectDir}`

File: tools/go-cli/build.gradle (hunk 1 above)
Comment (markusthoemmes) [inline tools/go-cli/build.gradle]: Agreeing with @mdeuser . OS detection should not be part of the buildstep unless it's specifically needed for the build itself.

File: tools/go-cli/build.gradle (hunk 2 above)
Comment (mdeuser) [inline tools/go-cli/build.gradle]: Quite right.  The comment did not keep up with the changes.  Copy destination is openwhisk/bin/go-cli

File: tools/go-cli/build.gradle (hunk 3 above)
Comment (mdeuser) [inline tools/go-cli/build.gradle]: @markusthoemmes - Thanks for taking a look.  Are you thinking of the groovy simple templating or something similar?  I like the readability as it is right now; but if there is some other benefit from templates I'm missing, would you please elaborate.

Comment (markusthoemmes) [inline tools/go-cli/build.gradle]: You can rewrite the line to 

`args "${projectDir}/../../bin/go-cli/${go_osname}/${go_osarch}/wsk", "${projectDir}/../../bin/go-cli/wsk", "-s"`

**Comment by csantanapr** [comment]
@mdeuser I think we should not use relative paths i.e. I had problems, I think is some problems bulding the correct paths
git clone openwhisk
cd openwhisk
`gradle tools:go-cli:distDocker`
```
:tools:go-cli:distBinary
Executing 'docker rm -f go-cli'
Executing 'mkdir /Users/csantanapr/Documents/dev/whisk/git/openwhisk/tools/go-cli/../../bin/go-cli/'
Executing 'docker run --name go-cli whisk/go-cli:latest'
Executing 'docker cp go-cli:/src/github.com/go-whisk-cli/build/. /Users/csantanapr/Documents/dev/whisk/git/openwhisk/tools/go-cli/../../bin/go-cli'
ln: -s: No such file or directory
:tools:go-cli:distBinary FAILED

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/csantanapr/Documents/dev/whisk/git/openwhisk/tools/go-cli/build.gradle' line: 29

* What went wrong:
Execution failed for task ':tools:go-cli:distBinary'.
> Process 'command 'ln'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 3 mins 26.897 secs
```

**Comment by csantanapr** [comment]
@mdeuser binaries get generated:

```
~/dev/whisk/git/openwhisk (pr-768)
🐑  $ ls bin/go-cli/
linux   mac     windows
~/dev/whisk/git/openwhisk (pr-768)
🐑  $ ls bin/go-cli/mac/
386   amd64 arm   arm64
~/dev/whisk/git/openwhisk (pr-768)
🐑  $ ./bin/go-cli/mac/amd64/wsk -h
__          ___     _     _
\ \        / / |   (_)   | |
 \ \  /\  / /| |__  _ ___| | __
  \ \/  \/ / | '_ \| / __| |/ /
   \  /\  /  | | | | \__ \   <
    \/  \/   |_| |_|_|___/_|\_\

Usage:
  wsk [command]

Available Commands:
  action      work with actions
  activation  work with activations
  package     work with packages
  rule        work with rules
  trigger     work with triggers
  sdk         work with the sdk
  property    work with whisk properties
  namespace   work with namespaces
  list        list triggers, actions, and rules in the registry for a namespace

Flags:
      --apihost string      whisk API host
      --apiversion string   whisk API version
  -u, --auth string         authorization key
  -d, --debug               debug level output
  -i, --insecure            bypass certificate checking
  -v, --verbose             verbose output

Use "wsk [command] --help" for more information about a command.
```

**Comment by dubeejw** [comment]
@mdeuser, looks like the symlink failed in the output Carlos posted. @csantanapr, we need the symlink to point to the Mac binary.

**Comment by mdeuser** [comment]
Unlike Linux, Mac `ln` requires a canonical path.  Commit with fix is on the way

**Comment by mdeuser** [comment]
@csantanapr - Would you kindly try out this fix again on your Mac.  Thanks!

**Comment by dubeejw** [comment]
Kicking build process off again. Looks like a random failure occurred.

**Comment by mdeuser** [comment]
@csantanapr - One last check from you and this PR is ready to merge... finally!

**Comment by mdeuser** [comment]
Carlos verified fix locally on his Mac; this PR is ready to merge
https://github.com/openwhisk/openwhisk/pull/590#issuecomment-228885588
/cc @csantanapr

**Comment by csantanapr** [comment]
@mdeuser sorry I put the results in the issue and not the PR 😫

**Comment by lfzCarlosC** [comment]
I can reproduce the issue when running hello case on  windows platform.

git clone --depth=1 https://github.com/openwhisk/openwhisk.git
cd openwhisk/tools/vagrant
./hello

then terminal throws error  csantanapr commented on June 27

**Comment by csantanapr** [comment]
Hi @lfzCarlosC this PR was merged a while ago. 
Nothing have change in the area of detecting Windows recently. 

Could you try again, open a new issue and include the error output?
There were some build issues yesterday around go CLI but they are fixed now.

**Comment by lfzCarlosC** [comment]
yes,I am considering filing a bug now

**Pull request merged**